Fix mesh corruption of CSG by using elalish/manifold

fire opened the pull request:
Fixes: https://github.com/godotengine/godot-proposals/issues/9711
Fixes: https://github.com/elalish/manifold/issues/91

The goal is to supersede https://github.com/godotengine/godot/pull/91748.

Work was done as part of the [V-Sekai development community](https://github.com/V-Sekai), which is focused on providing social VR functionality for the open-source Godot Engine.

- [x] Remove exceptions
- [x] Remove thrust
- [x] Remove glm
- [x] Singlethread is on
- [x] ~~Materials are not transferring~~

~~Ok to review for style. Trying to debug why the materials are broken.~~

Fixes: https://github.com/godotengine/godot/issues/43755
Fixes: https://github.com/godotengine/godot/issues/58637
Fixes: https://github.com/godotengine/godot/issues/41140

## Future work

- [ ] Move calls to CSG manifold to https://docs.godotengine.org/en/stable/classes/class_workerthreadpool.html

fire committed:
Add csg boolean operators using elalish/manifold.

Uses MeshGL64 for more floating point precision.

Co-Authored-By: 31 <31eee384@gmail.com>
Co-Authored-By: Claudio Z <120678869+cloudofoz@users.noreply.github.com>

31 commented:
Thanks for the ping on Discord about this. I finally got a chance to take a look. 🙂

I got materials working (it seems), but all I did was put my old code from https://github.com/godotengine/godot/pull/91748#pullrequestreview-2058765811 back in and change some names to fit the new place where this implementation lives: https://github.com/31/godot/commit/0338b1f3130433e76cb3b45a4217073ef346893c.


Admittedly, my testing is still quite limited. Here's a simple scene that is looking good to me: https://gist.github.com/31/a3e597b32ab31b132ddbc488d8eb0324. The stress-test-ish scene with text also still seems to be working fine: https://gist.github.com/31/c3c4bace42d4ca550ae6badf0f2b7cf9. (Maybe even faster? Been a while.)

---

I like the change of where this logic lives (vs. https://github.com/godotengine/godot/pull/91748). However, I'm not following why the implementation details were changed. It seems significantly more complex and I can't spot what it's fixing. Do you have a test case that this helps with?

fire commented:
As far as I know the implementation details changed because Mesh became MeshGL in manifold. The rest of the changes were a result of this. I didn't like that fact that we are passing an increasingly growing material dictionary so that was removed, but I caused an error.

Thanks for fixing!

fire commented:
The developers working at manifold works hard to remove exceptions and thrust, so we were able to vastly reduce the dependencies. The changes here were due to api updates.

fire commented on modules/csg/csg.h:
```diff
@@ -152,7 +158,7 @@ struct CSGBrushOperation {
 		HashMap<Vector3, int> vertex_map;
 		OAHashMap<VertexKey, int, VertexKeyHash> snap_cache;
-		float vertex_snap = 0.0;
+		float vertex_snap = 0;
```
```suggestion
		float vertex_snap = 0.0;
```

Restore old.

elalish commented:
Exciting!

Calinou reviewed (commented):
Tested locally, it works as expected.

Performance-wise, I can't really notice a difference when dragging around child CSG nodes (or the Path3D node in the racetrack at the end) in the [3D CSG](https://github.com/godotengine/godot-demo-projects/tree/master/3d/csg) demo.

Binary size comparison:

```
123,260,560 godot.linuxbsd.editor.x86_64
123,448,976 godot.linuxbsd.editor.x86_64.master
```

Once again, I'm surprised that the binary of this PR is 188 KB **smaller** (Linux x86_64 editor build with `optimize=speed lto=full`). I would expect it to be larger since several third-party libraries were added.

fire commented on modules/csg/csg.cpp:
```diff
@@ -278,188 +273,181 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
+void CSGBrushOperation::merge_brushes(Operation p_operation, const CSGBrush &p_brush_a, const CSGBrush &p_brush_b, CSGBrush &r_merged_brush, float p_vertex_snap) {
+	HashMap<int32_t, Ref<Material>> mesh_materials;
+	manifold::Manifold brush_a;
+	pack_manifold(&p_brush_a, brush_a, mesh_materials, p_vertex_snap / 1e3);
```
Any idea where this magic 1e3 came from?

fire replied:
I'll probably remove since i can't figure out if the divide by 1e3 was a bandaid for a different problem in an earlier iteration.

31 replied:
Well, another place something like 1e3 used to show up before I removed it was `std::vector<float> propertyTolerance(MANIFOLD_MAX, CLAMP(p_snap, 1e-3, INFINITY));`. (The PR just uses `mesh.precision` now rather than passing `propertyTolerance` to the `manifold::Manifold` constructor.)

31 commented on modules/csg/csg.cpp:
```diff
@@ -278,188 +273,181 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
+		const HashMap<int32_t, Ref<Material>> &mesh_materials,
+		CSGBrush *r_mesh_merge) {
+	Ref<StandardMaterial3D> default_material;
+	default_material.instantiate();
```
Apologies for this... I added it a while back, and it seems unlikely that it's a good idea to instantiate this on each call. There's no reason for it to be unique, AFAIK. 😄 However, I'm not really familiar with an existing shared default or where it would be correct to set one up.

On the other hand: maybe there should never need to be a default? Maybe every shape should have a material before we even get here? Or maybe null is (or should be) ok?

fire replied:
The only requirement seems to be that Ref<Material>() aka null has a unique slot, but it might be useful to have each Ref<Material>() be the same material rather than uncomparable.

fire replied:
I removed the default material.

cloudofoz commented:
Hi, this build crashes using my code:

`csg_source_node._update_shape()`
`var mesh_arrays = csg_source_node.get_meshes()`

I am calling **`_update_shape()`** to get the mesh data from the CSG node.
 [(https://github.com/godotengine/godot/issues/72814)](https://github.com/godotengine/godot/issues/72814)

But it works without problems with **v4.3-rc1_win64** and **v4.2.2-stable_win64**.

[backtrace.txt](https://github.com/user-attachments/files/16447690/backtrace.txt)

cloudofoz commented:
Just some more details.
The crash happens inside `static void pack_manifold(...)` during the construction of the manifold mesh:

**`r_manifold = manifold::Manifold(mesh);`**

In my test I am trying to intersect a mesh made with Blender (A) with a procedural mesh created in Godot via GDScript (B).

`pack_manifold` succeeds for the mesh A but it fails for the mesh B, due to an out-of-bounds index in the following code, where `x` becomes bigger than `parents` size:

```
I find(I x) {
  while (parents[x] != x) {
    parents[x] = parents[parents[x]];
    x = parents[x];
  }
  return x;
}
```

PiCode9560 commented:
> Hi, this build crashes using my code:
> 
> `csg_source_node._update_shape()`
> `var mesh_arrays = csg_source_node.get_meshes()`

While I tried to reproduce this, I came across this other way to crash the editor:


https://github.com/user-attachments/assets/a2657760-4d1f-4293-a2e8-dee0ae8f24a5





To do it, just add 3 CSGBox as children of the others. Rotate them around to make it messy. Lastly, move around the last CSGbox.

fire commented:
The trace seems a lot like https://github.com/elalish/manifold/issues/878

PiCode9560 commented:
Are all of these crashes need to be fixed before merging this PR?

cloudofoz commented:
I tried to change this line inside `pack_manifold(...)` :

`mesh.Merge();`

to:
```
/**
 * MeshGL::merge(): updates the mergeFromVert and mergeToVert vectors in order to create a
 * manifold solid. If the MeshGL is already manifold, no change will occur and
 * the function will return false. */
if (mesh.Merge()) {

	std::vector<int32_t> index_map(mesh.vertProperties.size() / MANIFOLD_MAX, -1);

	const auto vertices_count = mesh.mergeFromVert.size();

	for (auto i = 0; i < vertices_count; ++i) {

		index_map[mesh.mergeFromVert[i]] = mesh.mergeToVert[i];

	}
	const auto indices_count = mesh.triVerts.size();

	for (auto i = 0; i < indices_count; ++i) {

		if (index_map[i] > -1) {

			mesh.triVerts[i] = index_map[i];

		}
	}
}
```

My test works now and it doesn't crash anymore (not as "good" as the previous implementation at the moment, but it was just an experiment).

Please be kind because it's the first time I'm reading the manifold and godot source code, so I don't really know what I'm doing 😄

cloudofoz commented:
Here's the video:

https://github.com/user-attachments/assets/1fcf5437-54e7-4e87-8496-156e7b4027db

PiCode9560 commented:
@cloudofoz Your solution also fixed the crash that I posted earlier. 🎉

cloudofoz commented:
> @cloudofoz Your solution also fixed the crash that I posted earlier. 🎉

Glad of have been of help

fire commented:
I pushed the changes mentioned by @cloudofoz thanks!

fire commented:
Noted some crashes I'll try to get some reproduction steps for.

PiCode9560 commented:
I found a crash.
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.3.beta.custom_build (52ed05f413673170642d5ab2c01334496e793e4d)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] `anonymous namespace'::DedupePropVerts (C:\godot engine csg\godot\thirdparty\manifold\src\manif)
[1] manifold::Manifold::Impl::CreateFaces (C:\godot engine csg\godot\thirdparty\manifold\src\manifo)
[2] manifold::Manifold::Impl::Impl (C:\godot engine csg\godot\thirdparty\manifold\src\manifold\src\)
[3] manifold::Manifold::Impl::Impl (C:\godot engine csg\godot\thirdparty\manifold\src\manifold\src\)
[4] manifold::Manifold::Manifold (C:\godot engine csg\godot\thirdparty\manifold\src\manifold\src\ma)
[5] pack_manifold (C:\godot engine csg\godot\modules\csg\csg.cpp:386)
[6] CSGBrushOperation::merge_brushes (C:\godot engine csg\godot\modules\csg\csg.cpp:463)
[7] CSGShape3D::_get_brush (C:\godot engine csg\godot\modules\csg\csg_shape.cpp:215)
[8] CSGShape3D::_get_brush (C:\godot engine csg\godot\modules\csg\csg_shape.cpp:188)
[9] CSGShape3D::_update_shape (C:\godot engine csg\godot\modules\csg\csg_shape.cpp:307)
[10] CallableCustomMethodPointer<CSGShape3D>::call (C:\godot engine csg\godot\core\object\callable_)
[11] Callable::callp (C:\godot engine csg\godot\core\variant\callable.cpp:58)
[12] CallQueue::_call_function (C:\godot engine csg\godot\core\object\message_queue.cpp:221)
[13] CallQueue::flush (C:\godot engine csg\godot\core\object\message_queue.cpp:284)
[14] SceneTree::physics_process (C:\godot engine csg\godot\scene\main\scene_tree.cpp:494)
[15] Main::iteration (C:\godot engine csg\godot\main\main.cpp:4063)
[16] OS_Windows::run (C:\godot engine csg\godot\platform\windows\os_windows.cpp:1686)
[17] widechar_main (C:\godot engine csg\godot\platform\windows\godot_windows.cpp:181)
[18] _main (C:\godot engine csg\godot\platform\windows\godot_windows.cpp:208)
[19] main (C:\godot engine csg\godot\platform\windows\godot_windows.cpp:220)
[20] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[21] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================
```

To reproduce it. Set up the scene the [same way as I showed before](https://github.com/godotengine/godot/pull/94321#issuecomment-2265731584). And then the deepest child, resize one of the axis to 0 using the gizmo (set from the inspector doesn't crash it)

31 commented:
I got a crashing scene that repros upon load: https://gist.github.com/31/98da7abd2b0f52775e968423b7d513c6.
v4.4.dev.custom_build [9f7ce11cb]
(I used a text mesh to find the crash by adding one char at a time, then modified the tscn file to include the crashing char.)

fire commented:
Will investigate, but I queued up a lot of work so things are slow. Would appreciate help debugging.

31 commented:
I rebuilt on latest (v4.4.dev.custom_build [e0697ae49]), and a build failure forced me to update from MinGW 13.2.0-rt_v11-rev1 to 14.2.0-rt_v12-rev0, and now I can't get a crash anymore with text add or subtract. 🙂 However, I did pretty quickly find some bad meshes while subtracting somewhat longer text: https://gist.github.com/31/f12947d2e3abad5bc6ec0a3a9fa12611 🙁 Hopefully we aren't at the point where floating point and whatever build optimizations are getting in the way of reproducibility.

fire commented:
So the next steps would be to get a stack trace and try to see if the problem is in manifold or godot engine. 

If in manifold, we need to construct a upstream test case.

31 commented:
I was wondering about that. I think it might make sense to build into the editor a feature that exports all the Manifold inputs in a scene into a format that can be used as a "pure" Manifold test case. Not a trivial amount of work, to be sure, though.

It would help Manifold improve to have Godot's user base able to diagnose and report these things in complex scenes. Since the output is (usually) bad geometry and not crashes, and I don't think it's necessarily possible to automatically detect bad geometry, human eyes on this system seems like the real, practical way to help Manifold improve.

(Assuming Manifold isn't perfect already and this isn't all on us. 🙂)

elalish commented:
I'm not going to claim perfection! 😁 And yes, automatic generation of test cases would be amazing. In fact I built that same concept into Manifold for generating polygon test cases for our triangulator, since that's the source of most of our bugs.

fire commented:
@elalish is there a generate test recreation report button in manifold? :D Or where did you implemented it?

elalish commented:
The only one we have currently is if you compile with the `MANIFOLD_DEBUG` flag then it will test each polygon triangulation for geometric correctness and output a test case on the command line if it fails. However, if you're seeing a crash, that's likely a different problem. In any case, I think @31's suggestion is a good one.

fire commented:
Horray Manifold recently replaced the GLM dependency with linalg.h https://github.com/elalish/manifold/pull/984

fire commented:
Looking for team approval and style approval.

AThousandShips commented on modules/csg/csg.cpp:
```diff
@@ -278,188 +273,195 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
+		const CSGBrush *const p_mesh_merge,
+		manifold::Manifold &r_manifold,
+		HashMap<int32_t, Ref<Material>> &mesh_materials,
+		const float p_snap) {
```
```suggestion
		float p_snap) {
```
No `const` for simple types

smix8 commented:
Tested this PR a little and while it in general works fine I had a few random instances where for whatever reason certain parts of the mesh would turn out with wrong normals or smoothing.

It would look like this here with the spere mesh that touches the box mesh with union.

![csg_bug](https://github.com/user-attachments/assets/29fe96e1-4040-4e1a-b333-a1b62b5143f7)

When I change csg properties or move the objects a little it usually disappears. It is hard to reproduce because when I undo all the the changes and have the csg at the original bug state it does not show the issues again. When I try to save the scene to create an MRP it does not show the bug on loading. Any guess what could cause this?

fire commented:
I'll look through the code for normals modifying code.

I found:

```C++
face.smooth = mesh.vertProperties[property_i * mesh.numProp + MANIFOLD_PROPERTY_SMOOTH_GROUP] > 0.5f;
```

It could also be non deterministic when we merge on the edges of the two manifold shapes.

@elalish Might have more ideas.

fire commented:
@31 Any idea why this is off by one?

```C++
for (size_t run_i = 0; run_i < mesh.runIndex.size() - 1; run_i++) {
```

elalish commented:
It's off by one because our runIndex has an extra element to store the total size. I'd recommend using `mesh.runOriginalID.size()` instead. This way you can safely do things like: `run = {mesh.runIndex[run_i], mesh.runIndex[run_i + 1]}`.

elalish commented:
As for `vertProperties`, I'd need some more details on what kind of data you're trying to store/extract to help. First guess is you need to remember to duplicate any property vertices that you want to have different values on different triangles. And remember that the results you get will have those duplicates as well. If you are trying to recover a property that isn't really per-vert but a group of faces, I'd recommend using `runOriginalID` instead of `vertProperties`.

fire commented:
The deletion of the "default_material" seemed to cause a problem. Testing now with the default material restored.

fire commented:
Perhaps the previous theory is wrong.

It seems to be related to the low polygon nature of the meshes.


https://github.com/user-attachments/assets/046d5fea-4b25-4077-8b44-b93c5f41b88f

fire commented:
> recover a property that isn't really per-vert but a group of faces, I'd recommend using runOriginalID instead of vertProperties.

`face.smooth` and `face.invert` are per face rather than per face vertex.

Didn't you make an example of runOriginalID a while ago? If so can you link it?

elalish commented:
>Didn't you make an example of runOriginalID a while ago? If so can you link it?

It's in this thread: https://github.com/godotengine/godot/pull/91748#discussion_r1602101377

fire commented:
The difference seems to be that the CSGSphere3D and CSGBox3D don't automatically calculate tangents, but CSGMesh3D does.

fire commented:
@smix8 The artifact where the shading changes when the mesh goes from touching to intersection lies from the that the below list doesn't calculate tangents from the faces. It is possible and would be highly wanted by others and me, but it could be a different feature request after this is merged.

* CSGBox3D
* CSGCylinder3D
* CSGPolygon3D
* CSGSphere3D
* CSGTorus3D

What do you think about approving this for now?

A workaround is to use CSGMesh3D which does calculate tangents or add a StandardMaterial and set it to vertex shading.

smix8 commented:
So you mean this is not a regression with this PR? Just something that was already the case in master with those CSG nodes and I just did not encounter it by luck? In that case, yes I think it would be fine to improve that later in another PR.

fire commented:
I can confirm it happens in Godot Engine 4.3.stable.


https://github.com/user-attachments/assets/c0af731f-e29f-45ca-a410-1cb3fb526300

smix8 commented:
When testing `CSGMesh3D` node with a `SphereMesh` there is a noticeable performance different between this PR and master using the same build options.

I mean both versions stutter with a Mesh resource but with this PR I can barely control the mouse while dragging. Is this performance difference expected? What is causing it?

fire commented:
Will need to go over the code with a profiler to see what is the slowdown, but it's late Friday today, so will be delayed for a bit.

smix8 reviewed (approved):
Encountered a few CSG issues while testing this but apparently they all exist in the current master and are not regressions from this PR.

E.g. specific node types had issues or dont work with certain ops which need to be looked at (e.g. hide the non-working options from the inspector if they can't be fixed), same with some performance differences.

Everything outside of that so far worked fine. I would be ok with pushing the other fixes to followup PRs to keep the ball rolling.

lyuma commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -844,7 +844,7 @@ void TextureStorage::texture_2d_layered_initialize(RID p_texture, const Vector<R
 	}
 
-	Texture texture;
+	Texture texture = {};
```
Texture does not have default initializers, which is bad form, but this should be a separate PR.

I don't think we should be adding a single = {} just in this one place since it doesn't improve the overall code quality.

If there is a bug that we're working around or fixing by doing this, I'd like to be explicit about what it is. I double checked all the class members without default initializers and they should all be getting initialized here, so I'm skeptical that this change is needed.

lyuma commented on modules/csg/csg.cpp:
```diff
@@ -278,188 +273,195 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
-		r_merged_brush.materials.write[E.value] = E.key;
+void CSGBrushOperation::merge_brushes(Operation p_operation, const CSGBrush &p_brush_a, const CSGBrush &p_brush_b, CSGBrush &r_merged_brush, float p_vertex_snap, Ref<Material> p_default_material) {
+	HashMap<int32_t, Ref<Material>> mesh_materials;
+	manifold::Manifold brush_a;
+	_pack_manifold(&p_brush_a, brush_a, mesh_materials, p_default_material, p_vertex_snap);
```
This is called every time the CSGShape3D is dirty, right?

My main feedback here is that I am worried about continually packing and unpacking the manifold for each operation within CSGShape3D::_get_brush() . this seems like it could be a performance bottleneck.

I'd think that the point of this that we can keep each CSGBrush in manifold form and then a merge operation is now as simple as + or ^ or -.

(and separately, we do unpack once per frame if it's dirty just to submit to the graphics API.)

fire replied:
1. `+` `^` are commutative 
2 `-` is not commutative

I did some cleanup but wasn't able to rewrite the expressions.

lyuma commented on modules/csg/csg.cpp:
```diff
@@ -278,188 +273,195 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
+		mesh.runIndex.push_back(mesh.triVerts.size());
+
+		// Associate the material with an ID.
+		uint32_t reserved_id = r_manifold.ReserveIDs(1);
```
This is a static function? If we do run this once per CSG operation, if I have 1000 CSG brushes at 60 Hz, it will overflow the static uint32_t after just 20 hours of gameplay. If it's once per face, this uint32_t could overflow within minutes.

Is an overflow of the static ID generator expected / normal?

Is there a way to manage the allocation ourselves? I'm just feeling a bit uneasy about using a global static function even if it isn't hammered every frame.

fire replied:
are we generating 60 000 * 60 * 60 * 20 boolean operations in a nested tree if operations?

Also this might be an upstream limit until its changed. @elalish Not sure if doubles changed the run id limit

elalish replied:
I'd like to understand why you need to generate that many IDs. My first instinct is they're probably not being used as intended if you create that many. Their point is generally to be a way to map back to e.g. materials. I doubt you're creating that many unique materials, right?

fire replied:
@lyuma As far as I know we only generate one runid per material in a mesh. So I don't know how you are calculating this.

lyuma commented on modules/csg/csg.cpp:
```diff
@@ -278,188 +273,195 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
-			int inside_count = 0;
+	r_manifold = manifold::Manifold(mesh);
+	manifold::Manifold::Error err = r_manifold.Status();
+	if (err != manifold::Manifold::Error::NoError) {
+		print_error(String("Manifold creation from mesh failed:" + itos((int)err)));
+	}
```
Shouldn't this return a boolean or Error if it fails? Or is it impossible to trigger an error here (this is more like an assertion)

Might be good to use one of the ERR_ macros.

fire replied:
If the geometry is not manifold able I prefer it disappears rather to crash godot. An error message every frame is not fun either.

lyuma reviewed (commented):
I'll admit I don't know much about the manifold library itself, but the change seems to largely consolidate code around manifold packing and unpacking, which is good.

The thirdparty code addition is not too crazy, but that's more a question for maintainers to answer.

The main issue I'm worried about is that the process of converting to/from manifold looks at first glance to be a bit costly, and I'm worried it's doing the conversion too often.

fire commented:
@smix8 I fixed the 3d mesh normals.

![editor_screenshot_2024-11-07T181915](https://github.com/user-attachments/assets/b2e37233-c8ac-4b1b-b7f3-b89b0b5e9b75)
![editor_screenshot_2024-11-07T181952](https://github.com/user-attachments/assets/731db8a7-2086-45b3-be26-144ac30ab35f)

fire commented:
The lighting is wonky, and I don't know why.

lyuma reviewed (approved):
Looking good. The unpack optimization should help a lot.

There is still some room for optimization by caching the output of `_pack_manifold` for a given shape, transform and materials, but to decide if caching this makes sense, we would probably need to profile.

fire commented:
Ready to go.

fire commented:
Test cases for the optimization revert.

[csg_flicker.zip](https://github.com/user-attachments/files/17681543/csg_flicker.zip)

![Screenshot 2024-11-08 at 8 36 10 AM](https://github.com/user-attachments/assets/d2f63501-41b0-470d-91a9-0c30b0b06f9b)

![image](https://github.com/user-attachments/assets/7a25b387-a2e3-487d-80a4-a6cefdac48e1)

![editor_screenshot_2024-11-08T083835 copy](https://github.com/user-attachments/assets/12e7e347-7873-4cca-bf07-82254b82ffce)

fire commented:
Switched the code to lazily collect manifolds of the same type and if the operation changes we apply the collection with `manifold::Manifold::BatchBoolean(manifolds, current_op);`.

Moved the `_unpack_manifold` to the very end.

fire commented on thirdparty/README.md:
```diff
@@ -809,6 +820,16 @@ Files extracted from upstream source:
+- Version: master (1ffbc6f884ea1da89e104a5996cf8a726db673d5, 2023)
+- License: Public Domain
+
+Files extraced from upstream sources:
```
```suggestion
Files extracted from upstream sources:
```

aaronfranke commented on modules/csg/csg_shape.h:
```diff
@@ -51,6 +51,7 @@ class CSGShape3D : public GeometryInstance3D {
 	};
 
 private:
+	Ref<Material> default_material = memnew(StandardMaterial3D);
```
I don't think this is how `Ref<>` types are instantiated. It's usually like `default_material.instantiate();`.

aaronfranke commented on modules/csg/csg_shape.cpp:
```diff
@@ -167,78 +169,244 @@ void CSGShape3D::_make_dirty(bool p_parent_removing) {
-	if (dirty) {
-		if (brush) {
-			memdelete(brush);
+enum {
```
I don't like unnamed enums. We should rename this to `ManifoldProperty`.

```suggestion
enum ManifoldProperty {
```

I also tried changing this to `enum class` but since these enum values are used as an integers in most places that would be worse in this case.

aaronfranke reviewed (commented):
@fire asked me to review this PR. First up, testing to see if it works:

* https://github.com/godotengine/godot/issues/58637
    * The test project was made with Godot 4.0 so it may be outdated.
    * Cannot reproduce the bug in master, it looks visually correct already. However, moving the CSG box causes the editor to freeze and stop responding.
    * With this PR, there is still no bug, and the editor doesn't freeze anymore. ✅
* https://github.com/godotengine/godot/issues/43755
    * The test project is made with Godot 3.x so it had to be converted to Godot 4.x.
    * I am able to reproduce the bug in master.
    * With this PR, the bug is fixed. ✅
* https://github.com/godotengine/godot/issues/41140
    * The MRP is a scene file, not a project, but it seems to load fine in master.
    * I am able to reproduce the bug in master.
    * With this PR, the bug is fixed. ✅ Plus the topology looks cleaner overall. ✅
    * In both master and this PR, the outline of the CSG geometry becomes outdated when moving meshes. Clicking on the mesh with outdated outlines causes the outlines to refresh.
* All test cases:
    * In both master and this PR, there is massive Z-fighting when a CSG box is selected.
    * I did not test materials, the MRPs did not include any.

20 thousand new lines of code is a lot, unfortunately, but @fire has assured me that this is as cut down as it can be. It makes sense, calculating CSG manifolds is not a simple operation. All of this 'bloat' can be disabled by disabling the CSG module, so it's a non-issue for people making a minimal build of Godot. Plus, it removes a lot of broken first-party code that we don't have to maintain anymore, replacing it with working third-party code, which is great.

As for a code review, I'm not an expert in the CSG module so I cannot give it an in-depth review. However, it looks pretty good overall, it's all self-contained in the CSG module, but I have two notes:

fire commented:
Quick-hull is removable in a newer manifold git revision to reduce the code size, but @31 wanted to make a hull operator on top of this pr.

Updating to a newer revision is blocked on https://github.com/elalish/manifold/issues/1027

fire commented on modules/csg/csg_shape.cpp:
```diff
@@ -167,78 +169,244 @@ void CSGShape3D::_make_dirty(bool p_parent_removing) {
+	MANIFOLD_MAX
+};
+
+static void _unpack_manifold(
```
I still don't like bare structures, but had some difficulty with includes and symbols moving it to a class.

Suggestions?

aaronfranke replied:
I like bare methods in the .cpp file, for things that don't need member access or being called from other files, this is quite nice. It's like the ultimate version of `private`.

elalish commented:
>Quick-hull is removable in a newer manifold git revision to reduce the code size, but @31 wanted to make a hull operator on top of this pr. https://github.com/godotengine/godot-proposals/issues/9933

Not so much removable as integrated. We removed the quick-hull dependency in favor of making it a native component with quite a bit less code. So you should be good to go updating and still building your hull operator on top.

fire commented:
Renamed `MANIFOLD_MAX` to `MANIFOLD_PROPERTY_MAX`.

akien-mga commented:
@BastiaanOlij I remember you doing quite some experiments with CSG a few years back. If you still have some of those projects laying around, it would be great if you could test/review this overhaul.

I started doing a code review myself, will finish next week.

BastiaanOlij commented:
Well, I've extensively used CSG to create tracks in this project: https://github.com/BastiaanOlij/vehicle-demo/pull/15

But compiling this branch and loading this project I'm confronted with a frozen UI and a log full of `Moving child node positions inside the SceneTree is only allowed from the main thread. Use call_deferred("move_child, child, index).` errors. This is likely unrelated to this PR though.
Also I don't believe I use threading in this project other than using the ResourceLoader to background load this project.

The project stalls further with an unending "Object was deleted while awaiting a callback" error though only output to console. Again I doubt this is related to this PR itself.

![image](https://github.com/user-attachments/assets/5c41365a-dad6-4dfd-be37-2d1f5c4957d1)
Without being able to move my camera it's hard to judge but at first glance, I think all the UVs are off.

I'll see if I can run this on master and see if the same problems arise or if there are issues specific to this PR.

fire commented:
Can you describe the messed up UVs?

fire commented:
Can confirm messed up uvs.

# CSG Branch

![image](https://github.com/user-attachments/assets/ec1be50d-091e-4a15-aeb9-4c2062c5b7c9)

# Godot 4.3 from Steam on macos

![image](https://github.com/user-attachments/assets/6caa13dc-9de5-4878-98e7-b5884f33ef96)

BastiaanOlij commented:
Ok, I don't know why but the issue with threading has mysteriously disappeared after I did a bunch of work trying to track it down.. 

So indeed, just the UV issue remains. Looks like CSG polygons are loosing the UVs when they are set to tile along the track

fire commented:
Can you clear your .godot cache and let me know if it makes a difference?

# 4.4 with csg

![image](https://github.com/user-attachments/assets/2fb1d019-c873-4681-8d95-aa711a300669)

BastiaanOlij commented:
@fire seems to make some difference but I can see at the starting line that my normal map is only correct only in one direction. The CSG is set to close the loop so I think that it doesn't create the last segment properly (probably just reuses the vertices at the start ending up with the wrong UVs):

![image](https://github.com/user-attachments/assets/8f6b7acb-9eb2-42b1-b7a0-8ef6ec84a46d)

The pitlane is completely wrong for me:
![image](https://github.com/user-attachments/assets/27815ba8-d5bf-40c7-9a1a-70f34a753c81)

BastiaanOlij commented:
Ah, I'm correct it messes up at the closing segment:
![image](https://github.com/user-attachments/assets/676427fc-c22f-4542-9614-d41500391458)

The track is correct until the last segment.

fire commented:
I will upgrade to the latest master and use the closed property of Curve3d https://github.com/godotengine/godot/pull/86195.

fire commented:
If I used the closed property of path there's some z-fighting 
![image](https://github.com/user-attachments/assets/0b11fc6d-eaa0-41f3-8ad5-b60c02fa8182)

BastiaanOlij commented:
The close on the Curve3D doesn't align the control  handles, it just places the end point on the same location as the start point. That doesn't work well with extrusions.

fire commented:
@elalish, I'm stumped. Do you have any ideas?

fire commented:
@BastiaanOlij @elalish, this isn't using the mesh booleans; as far as I can tell, it is only the MeshGL data structure. In csgpolygon3d, we manually create a triangle mesh, which feels suspect.

Also, the final result is 1. inverted 2. the shader uses UV = Vector2(uv.y, uv.x) * Vector2(8.0, 0.1)

akien-mga commented on modules/csg/csg.cpp:
```diff
@@ -33,167 +33,13 @@
-
-	return max_separation2 < p_vertex_snap2;
-}
+#include "thirdparty/manifold/include/manifold/manifold.h"
```
Adding `thirdparty_dir + "include"` to the `CPPPATH` is so that you can do this:
```suggestion
#include <manifold/manifold.h>
```

May or may not be relevant for manifold in the future, but that's also the syntax which would work if we supported building against a system packaged manifold library on Linux. (Again here that's highly theoretical, I don't suggest working on adding support for that - I don't see any Linux distro packaging it.)

akien-mga commented on modules/csg/csg_shape.cpp:
```diff
@@ -31,6 +31,8 @@
 #include "csg_shape.h"
 
 #include "core/math/geometry_2d.h"
+#include "manifold/common.h"
+#include "manifold/manifold.h"
```
The explicit include for `manifold/common.h` could be removed too as it's already included by `manifold.h`.
Usually libraries like this are designed so that for the basic API there's a single header to include named like the library, and it comes with batteries included.
```suggestion
#include "core/math/geometry_2d.h"

#include <manifold/common.h>
#include <manifold/manifold.h>
```

akien-mga commented on thirdparty/README.md:
```diff
@@ -546,6 +546,17 @@ Patch `godot-node-debug-fix.patch` workarounds shadowing of Godot's Node class
+
+- `src/`
+- `AUTHORS`, `LICENSE`
+
```
Two empty lines between sections.
```suggestion


```

akien-mga commented on thirdparty/README.md:
```diff
@@ -546,6 +546,17 @@ Patch `godot-node-debug-fix.patch` workarounds shadowing of Godot's Node class
+## manifold
+
+- Upstream: https://github.com/elalish/manifold
+- Version: master (12b43f50c44988cf3bc650df9fda180c8cc06163, 2024)
```
```suggestion
- Version: git (12b43f50c44988cf3bc650df9fda180c8cc06163, 2024)
```

But if that's the commit we're tracking, it seems to be a bit older than the newly released 3.0.0, so I would suggest updating the lib to https://github.com/elalish/manifold/releases/tag/v3.0.0 and then documenting it like this:
```suggestion
- Version: 3.0.0 (5d127e57fbfb89225a8e905d0d914ccc86c139c8, 2024)
```

akien-mga commented on modules/csg/SCsub:
```diff
@@ -6,7 +6,43 @@ Import("env_modules")
+
+thirdparty_dir = "#thirdparty/manifold/"
+thirdparty_sources = [
+    "src/cross_section/cross_section.cpp",
```
The upstream CMakeLists.txt flags this as optional. Do we use it?
I don't see a reference to `cross_section.h` in our code nor in manifold code, so presumably `src/cross_section/` and `include/cross_section.h` can be removed.

akien-mga replied:
`include/meshIO.h` can also be removed as we're not including its `.cpp`.

akien-mga commented on modules/csg/SCsub:
```diff
@@ -6,7 +6,43 @@ Import("env_modules")
+thirdparty_dir = "#thirdparty/manifold/"
+thirdparty_sources = [
+    "src/cross_section/cross_section.cpp",
+    "src/constructors.cpp",
+    "src/polygon.cpp",
+    "src/edge_op.cpp",
+    "src/face_op.cpp",
+    "src/impl.cpp",
+    "src/boolean_result.cpp",
+    "src/boolean3.cpp",
+    "src/manifold.cpp",
+    "src/properties.cpp",
+    "src/smoothing.cpp",
+    "src/sort.cpp",
+    "src/csg_tree.cpp",
+    "src/subdivision.cpp",
+    "src/quickhull.cpp",
```
What logic was used for the order of the files? Usually we'd order alphabetically, or in the same order as upstream's buildsystem lists them, which here seems alphabetical too:

https://github.com/elalish/manifold/blob/master/src/CMakeLists.txt#L15-L31

fire replied:
It's in random order. How can I sort?

AThousandShips replied:
Most code editors come with a sort lines function

akien-mga reviewed (commented):
Only reviewed the way thirdparty code is included, it seems overall pretty good to me. Just left a few nitpicks.

For the actual CSG functionality, I'm a bit too remote from it to properly assess the change, so I'd just say `needs testing` (which can happen post-merge in dev snapshots, but let's first get to the bottom of the issue identified by Bastiaan).

akien-mga commented:
Ah and I forgot, this should be added too:
```diff
diff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index 2ece06eebb..7f9a23c931 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -334,6 +334,11 @@ Comment: WebP codec
 Copyright: 2010, Google Inc.
 License: BSD-3-clause
 
+Files: ./thirdparty/manifold/
+Comment: Manifold
+Copyright: 2020-2024, The Manifold Authors
+License: Apache-2.0
+
 Files: ./thirdparty/mbedtls/
 Comment: Mbed TLS
 Copyright: The Mbed TLS Contributors
```

fire commented:
I made a minimal bug recreation project.

[csg-error-mrp 2.zip](https://github.com/user-attachments/files/17912469/csg-error-mrp.2.zip)

![image](https://github.com/user-attachments/assets/85412c2d-55ba-442b-9c9d-b776042181ba)

fire commented:
![image](https://github.com/user-attachments/assets/2d7b3850-dc44-4abc-92ba-7b09c24e7a31)

![image](https://github.com/user-attachments/assets/8fc8916a-b5fd-418f-95aa-a020ff669f4d)

fire commented:
@BastiaanOlij and others. The code is ready for review again.

BastiaanOlij reviewed (approved):
I've mostly focussed on the CSGPolygon usage in my vehicle example, that seems to all be working fine now.

I did do a bit of testing combining various CSG shapes that I knew gave problems in the past and were handled very well now, but didn't do any major stress testing on that area.

Seeing how few maintainers we've had that really wanted to dive into the heart of the CSG logic, relying on a maintained thirdparty library seems like a solid choice.

AThousandShips commented on modules/csg/csg_shape.cpp:
```diff
@@ -31,6 +31,7 @@
 #include "csg_shape.h"
 
 #include "core/math/geometry_2d.h"
+#include <manifold/manifold.h>
```
```suggestion

#include <manifold/manifold.h>
```
Include grouping

AThousandShips reviewed (approved):
Just a nitpick I missed last pass

akien-mga reviewed (approved):
Ready to merge!

zhangjiangen11 commented:
https://github.com/catprisbrey/Cats-Godot4-Modular-Souls-like-Template  load this project crash!

zhangjiangen11 commented:
After reverting this commit, the scene can be loaded normally.

fire commented:
@zhangjiangen11 Can you create a new issue? I suggest clearing .godot because I can't recreate the crash.

goatchurchprime commented:
Here's a quick stress test of it.  A few examples like this could feed into a very good blog post about it.
 ![Screenshot from 2024-12-02 12-55-09](https://github.com/user-attachments/assets/afae535b-85ba-447c-b925-486f2afc4060)

fire commented:
Can you post the script? The top portion of the sponger seems corrupted.

goatchurchprime commented:
Here it is.   [csgmanifoldtest.zip](https://github.com/user-attachments/files/17983282/csgmanifoldtest.zip)

Set `mengerdimension = 4` and it takes 2 minutes instead of 0.2 seconds.
Set `holelength = 1.001` to ensure that the holes go all the way through

The holes gaps on the rotated combiners of boxes might be misaligned if 90degree rotation is implemented using the axis vectors `(sin(1.5707963), cos(1.5707963))` instead of `(0,1)` (radians are evil).

I can't explain why one the triangles of the middle hexagram star on each edge is too large.

fire commented:
I modified the snap values, and it seemed to work ok.

![image](https://github.com/user-attachments/assets/e5b1f20b-0590-4780-9c61-5f816f8a90ba)

[csgmanifoldtest 2.zip](https://github.com/user-attachments/files/17983723/csgmanifoldtest.2.zip)

goatchurchprime commented:
Is `snap` implemented entirely differently in manifold?  

I think these bits are different when you compare to the online version.  No idea why.  
![image](https://github.com/user-attachments/assets/3329d6da-808d-4bf3-9e53-aec346c7a13f)

fire commented:
I associated godot engine `snap` with `tolerance`

elalish commented:
There's no snapping in manifold itself. It almost looks like it's using 0.33 instead of a 1/3 or something. @fire even yours looks quite asymmetric and wrong to me. 

>I associated godot engine snap with tolerance

That might be dangerous... I'm not sure exactly how snap is used internally by Godot, but we use tolerance for mesh simplification, which can yield pretty unexpected results if the value gets too large.

fire commented:
We can deprecate snap and use the default manifold tolerance. How does that sound?

elalish commented:
Yeah, that sounds good to me.

31 commented:
> We can deprecate snap and use the default manifold tolerance. How does that sound?

I already saw an issue with the default snap/tolerance a while back, example in this issue: https://github.com/godotengine/godot/issues/93240

I'm not sure how to rectify the meanings, but from what I've seen, it probably matters to expose it somehow. *Maybe* you can get around it by scaling/unscaling, but that doesn't sound all that nice to deal with for prototyping in-editor.

elalish commented:
@31 - the issue you linked to is an example of the same problem: the mesh being broken by choosing too large a "snap" tolerance value. I still think @fire's approach will be the most reliable solution for that class of problems as well, particularly if snap is removed entirely.

31 commented:
> the issue you linked to is an example of the same problem: the mesh being broken by choosing too large a "snap" tolerance value.

I understand that, but I don't have a sense for what the default Manifold tolerance behavior means. If I scaled the text 1/100, would the default behavior show a similar problem? 1/10000? Or is the default behavior close enough to floating point epsilon that it's difficult to even encounter that kind of problem?

I wasn't able to find any info in the Manifold docs that really told me what situation to expect here. 😕

elalish commented:
Yes, the default is floating point epsilon.

ydeltastar commented:
Looks like this also fixed https://github.com/godotengine/godot/issues/94174.
